core/state/snapshot/generate.go

```diff
@@ -159,169 +181,441 @@ func journalProgress(db ethdb.KeyValueWriter, marker []byte, stats *generatorSta
+// the error will be returned to abort the entire procedure.
+func (dl *diskLayer) proveRange(root common.Hash, prefix []byte, kind string, origin []byte, max int, valueConvertFn func([]byte) ([]byte, error)) (*proofResult, error) {
+	var (
+		keys    [][]byte
+		vals    [][]byte
+		proof   = rawdb.NewMemoryDatabase()
+		aborted = false
+	)
+	iter := dl.diskdb.NewIterator(prefix, origin)
+	defer iter.Release()
+
+	for iter.Next() {
+		key := iter.Key()
+		if len(key) != len(prefix)+common.HashLength {
+			continue
+		}
+		if len(keys) == max {
+			aborted = true
+			break
+		}
+		keys = append(keys, common.CopyBytes(key[len(prefix):]))
+
+		if valueConvertFn == nil {
+			vals = append(vals, common.CopyBytes(iter.Value()))
+			continue
+		}
+		val, err := valueConvertFn(iter.Value())
+		if err != nil {
+			// Special case, the state data is corrupted (invalid slim-format account),
+			// don't abort the entire procedure directly. Instead, let the fallback
+			// generation to heal the invalid data.
+			//
+			// Here append the original value to ensure that the number of key and
+			// value are the same.
+			vals = append(vals, common.CopyBytes(iter.Value()))
+			continue
+		}
+		vals = append(vals, val)
+	}
+	// The snap state is exhausted, pass the entire key/val set for verification
+	if origin == nil && !aborted {
+		stackTr := trie.NewStackTrie(nil)
+		for i, key := range keys {
+			stackTr.TryUpdate(key, common.CopyBytes(vals[i]))
+		}
+		if gotRoot := stackTr.Hash(); gotRoot != root {
+			return &proofResult{keys: keys, vals: vals, hasMoreElems: aborted, proofErr: errors.New("wrong root")}, nil
+		}
+		return &proofResult{keys: keys, vals: vals, hasMoreElems: aborted, proofErr: nil}, nil
+	}
+	// Snap state is chunked, generate edge proofs for verification.
+	tr, err := trie.New(root, dl.triedb)
+	if err != nil {
+		log.Error("Missing trie", "root", root, "err", err)
+		return nil, err
+	}
+	// Firstly find out the key of last iterated element.
+	var last []byte
+	if len(keys) > 0 {
+		last = keys[len(keys)-1]
+	}
+	// Generate the Merkle proofs for the first and last element
+	if origin == nil {
+		origin = common.Hash{}.Bytes()
+	}
+	if err := tr.Prove(origin, 0, proof); err != nil {
+		log.Debug("Failed to prove range", "kind", kind, "origin", origin, "err", err)
+		return &proofResult{keys: keys, vals: vals, hasMoreElems: aborted, proofErr: err, tr: tr}, nil
+	}
+	if last != nil {
+		if err := tr.Prove(last, 0, proof); err != nil {
+			log.Debug("Failed to prove range", "kind", kind, "last", last, "err", err)
+			return &proofResult{keys: keys, vals: vals, hasMoreElems: aborted, proofErr: err, tr: tr}, nil
+		}
+	}
+	// Verify the state segment with range prover, ensure that all flat states
+	// in this range correspond to merkle trie.
+	_, _, _, _, err = trie.VerifyRangeProof(root, origin, last, keys, vals, proof)
+	// Previously, we took the 'cont' from trie.VerifyRangeProof and used below.
```

Comment:
@holiman I think you are right here. Originally it's wrong. 

I think all the operations are based on the snap states in the disk. The `cont` flag should also refers the snaps states, instead of the trie.

Otherwise there is a corner case: the trie says nothing left, but in the database there is an extra state left and it won't be deleted.

So for whatever range, we should iterate all the states in the database and trie is only used for verification.